internal/linker/link_node_modules.ts

```diff
@@ -276,36 +297,72 @@ export async function main(args: string[], runfiles: Runfiles) {
   // Now add symlinks to each of our first-party packages so they appear under the node_modules tree
   const links = [];
 
-  const linkModule =
-      async (name: string, root: LinkerRoot, modulePath: string) => {
-    let target: string = '<package linking failed>';
-    switch (root) {
-      case 'bin':
-        // FIXME(#1196)
-        target = path.join(workspaceAbs, bin, toWorkspaceDir(modulePath));
-        break;
-      case 'src':
-        target = path.join(workspaceAbs, toWorkspaceDir(modulePath));
-        break;
-      case 'runfiles':
-        target = runfiles.resolve(modulePath) || '<runfiles resolution failed>';
-        break;
-    }
+  const linkModules =
+      async (modules: {name: string, root: LinkerRoot, modulePath: string}[]) => {
+    for (const m of modules) {
+      let target: string = '<package linking failed>';
+      switch (m.root) {
+        case 'bin':
+          // FIXME(#1196)
+          target = path.join(workspaceAbs, bin, toWorkspaceDir(m.modulePath));
+          break;
+        case 'src':
+          target = path.join(workspaceAbs, toWorkspaceDir(m.modulePath));
+          break;
+        case 'runfiles':
+          target = runfiles.resolve(m.modulePath) || '<runfiles resolution failed>';
+          break;
+      }
 
-    await symlink(target, name);
+      // ensure the subdirectories exist
+      mkdirp(path.dirname(m.name));
+      await symlink(target, m.name);
+    }
   }
 
-  for (const m of Object.keys(modules)) {
-    const segments = m.split('/');
-    if (segments.length > 2) {
-      throw new Error(`module ${m} has more than 2 segments which is not a valid node module name`);
-    }
-    if (segments.length == 2) {
-      // ensure the scope exists
-      mkdirp(segments[0]);
-    }
-    const [kind, modulePath] = modules[m];
-    links.push(linkModule(m, kind, modulePath));
+  // Group nested modules as these need to be symlinked in order.
+  // For example, given a list of module keys such as:
+  // ['a', '@foo/c/c/c/c', 'b/b', 'b', '@foo/c', '@foo/c/c']
+  // this reducer should output the groups list:
+  // [ [ '@foo/c', '@foo/c/c', '@foo/c/c/c/c' ], [ 'a' ], [ 'b', 'b/b' ] ]
+  const groupedModules = Object.keys(modules).sort().reduce(
```

Comment:
seems like it would be better to add unit tests for this function rather than testing all the conditions via the medium test that uses the filesystem

Reply:
that's a good idea; I'll refactor the function so its exported and add some unit tests for it

Reply:
broke the function out and added a unit test that makes it very clear what it is supposed to do :)

```
  it('should group and reduce modules appropriate', () => {
    const modules = {
      'a': ['src', `${workspace}/path/to/lib_a`],
      '@foo/c/c/c/c': ['bin', `${workspace}/path/to/foo_cccc`],
      'b/b': ['src', 'other_wksp/path/to/lib_bb'],
      'b': ['src', 'other_wksp/path/to/lib_b'],
      '@foo/c': ['bin', `${workspace}/path/to/foo_c`],
      '@foo/c/c': ['bin', `${workspace}/path/to/foo_cc`],
      '@foo/d/bar/fum/far': ['bin', `${workspace}/path/to/foo_d/bar/fum/far`],
      '@foo/d/bar': ['bin', `${workspace}/path/to/foo_d/bar`],
      // don't include `@foo/d` as the group & reduce function  should derive that
      // from the lowest common denominator of the module name & module path
    };

    const result = linker.groupAndReduceModules(modules);

    const expected = [
      [
        // `@foo/c`, `@foo/c/c` and `@foo/c/c/c/c` should
        // be grouped but not reduced as their arrangement
        // on disk does not match the modules names
        {'name': '@foo/c', 'root': 'bin', 'modulePath': `${workspace}/path/to/foo_c`},
        {'name': '@foo/c/c', 'root': 'bin', 'modulePath': `${workspace}/path/to/foo_cc`},
        {'name': '@foo/c/c/c/c', 'root': 'bin', 'modulePath': `${workspace}/path/to/foo_cccc`}
      ],
      [
        // `@foo/d/bar/fum/far` and `@foo/d/bar` should
        // be grouped and reduced to `@foo/d`
        {'name': '@foo/d', 'root': 'bin', 'modulePath': `${workspace}/path/to/foo_d`},
      ],
      [{'name': 'a', 'root': 'src', 'modulePath': `${workspace}/path/to/lib_a`}],
      [
        // `b`, `b/b` should be grouped but not reduced as
        // their arrangement on disk does not match the modules names
        {'name': 'b', 'root': 'src', 'modulePath': 'other_wksp/path/to/lib_b'},
        {'name': 'b/b', 'root': 'src', 'modulePath': 'other_wksp/path/to/lib_bb'}
      ]
    ];

    expect(JSON.stringify(result)).toEqual(JSON.stringify(expected));
  })
```